modules/chronograf-server/main.tf

```diff
@@ -55,5 +110,49 @@ resource "aws_security_group_rule" "allow_all_outbound" {
   protocol    = "-1"
   cidr_blocks = ["0.0.0.0/0"]
 
-  security_group_id = "${aws_security_group.chronograf_security_group.id}"
+  security_group_id = "${aws_security_group.lc_security_group.id}"
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+# ATTACH AN IAM ROLE TO EACH EC2 INSTANCE
+# We can use the IAM role to grant the instance IAM permissions so we can use the AWS CLI without having to figure out
+# how to get our secret AWS access keys onto the box. We export the ID of the IAM role as an output variable so users
+# can attach custom policies.
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "aws_iam_instance_profile" "instance_profile" {
+  name_prefix = "${var.cluster_name}"
+  path        = "${var.instance_profile_path}"
+  role        = "${aws_iam_role.instance_role.name}"
+
+  # aws_launch_configuration.launch_configuration in this module sets create_before_destroy to true, which means
+  # everything it depends on, including this resource, must set it as well, or you'll get cyclic dependency errors
+  # when you try to do a terraform destroy.
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+resource "aws_iam_role" "instance_role" {
+  name_prefix        = "${var.cluster_name}"
+  assume_role_policy = "${data.aws_iam_policy_document.instance_role.json}"
+
+  # aws_iam_instance_profile.instance_profile in this module sets create_before_destroy to true, which means
+  # everything it depends on, including this resource, must set it as well, or you'll get cyclic dependency errors
+  # when you try to do a terraform destroy.
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+data "aws_iam_policy_document" "instance_role" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
```

Comment:
Question, why does the autoscaling group need this?

Reply:
TBH, I've never really thought of this. It's something that's always been used so I just went with it. I'm working on the multi-cluster example next, so I'll experiment with this to know what exactly breaks if it's not present

Reply:
Not sure if this is the question you're asking, but some context:

IAM Roles contain two types of policies:

1. Trust relationships: this specifies who or what can assume the IAM policy. Above, we're grant EC2 Instances (type `Service`, identifier  `ec2.amazonaws.com`) the ability to assume this IAM Role. If this was an IAM Role for an ECS Task, we'd instead need to give ECS permissions to assume the role (type `Service`, identifier `ecs.amazonaws.com`).

1. IAM permissions: this specifies what permissions you have once you assume the role. It doesn't look like we're attaching any permissions to this particular role, but by creating the role and exporting its ID as an output variable, users of this module will be able to attach various IAM permissions to the role (e.g., permissions to write logs to CloudWatch) and the EC2 Instances in this ASG will be able to use those permissions.